### PR TITLE
rados-Backfill/Recovery: rados object that data size is 0 and this object have a large amount of omap key-value. when primary osd is backfilled, this object will be lost

### PR DESCRIPTION
**Pull request by alcyoneus86**
rados-Backfill/Recovery: rados object that data size is 0 and this object have a large amount of omap key-value. when primary osd is backfilled, this object will be lost. issues/53757
Signed-off-by: xingyu wang alcyoneus86@163.com

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by satoru-takeuchi** [comment]
Is this fix under review? Or has it been overlooked? This PR seems to be serious for RGW because bucket index objects are "rados object that data size is 0 and this object have a large amount of omap key-value.".

**Comment by satoru-takeuchi** [comment]
@alcyoneus86 Could you update this PR to pass the CI?

**Comment by neha-ojha** [comment]
jenkins retest this please

**Comment by ronen-fr** [comment]
@satoru-takeuchi : the commit signature seems to be in the wrong format ("wrong" here - not accepted by the
CI)

File: src/osd/ReplicatedBackend.cc [hunk 1]
```diff
@@ -2214,7 +2214,7 @@ bool ReplicatedBackend::handle_push_reply(
     PushInfo *pi = &pushing[soid][peer];
     bool error = pushing[soid].begin()->second.recovery_progress.error;
 
-    if (!pi->recovery_progress.data_complete && !error) {
+    if ((!pi->recovery_progress.data_complete || !pi->recovery_progress.omap_complete)&& !error) {
```
Comment (ronen-fr) [inline src/osd/ReplicatedBackend.cc:2217]: a missing blank before the '&&'

**Comment by satoru-takeuchi** [comment]
@ronen-fr 

> the commit signature seems to be in the wrong format ("wrong" here - not accepted by the
CI)

Yes, so I've asked @alcyoneus86 to update this PR. But he've not replied yet.

I consider that the bug to be fixed in this PR is serious. So if possible I'd like to make another PR having the same fix as this PR. Of course, signed-off-by will be his name. Does it make sense?

**Comment by ronen-fr** [comment]
> Yes, so I've asked @alcyoneus86 to update this PR. But he've not replied yet.
> 
> I consider that the bug to be fixed in this PR is serious. So if possible I'd like to make another PR having the same fix as this PR. Of course, signed-off-by will be his name. Does it make sense?

Maybe wait for the formal approval first, just in case there will be more issues suggested? your call

**Comment by satoru-takeuchi** [comment]
> Maybe wait for the formal approval first,

OK, I'll wait for a while.

**Comment by ronen-fr** [comment]
@satoru-takeuchi , @alcyoneus86 : I've now spent two full days trying to create the faulty scenario that is supposed to be fixed by this PR - with no success. It seems as though the code does make sure to never set data_complete before omap_complete
(and, btw, omap_complete is checked in ObjectRecoveryProgress.is_complete()).
Can you show me how to reproduce a problem?
Thanks

**Comment by ronen-fr** [comment]
jenkins retest this please

**Comment by satoru-takeuchi** [comment]
@alcyoneus86 Please tell us if you have a reproducer. In addition, I'd like to know how did you find this problem, e.g. when reading source code, or encounted this problem in real system, and so on.

@ronen-fr Unfortunately I've never reproduced this problem. I just thought that this problem seems to be serious as a result of looking at the commit description. Anyway, I'll also try to reproduce this problem.

**Comment by vumrao** [comment]
cc @vumrao

**Comment by jdurgin** [comment]
> @satoru-takeuchi , @alcyoneus86 : I've now spent two full days trying to create the faulty scenario that is supposed to be fixed by this PR - with no success. It seems as though the code does make sure to never set data_complete before omap_complete (and, btw, omap_complete is checked in ObjectRecoveryProgress.is_complete()). Can you show me how to reproduce a problem? Thanks

It looks like it is only possible with FileStore.

The condition here may be fulfilled if extents are removed from the list by readv():

https://github.com/ceph/ceph/blob/6b4382aac7596685c31e078ffaf404c1df196819/src/osd/ReplicatedBackend.cc#L2143-L2148

The readv implementation for FileStore (using the default one in ObjectStore.h) will remove extents that have no data. BlueStore has its own readv implementation which does not modify the interval_set of extents passed in though, so it would only be possible with FileStore using certain allocation patterns and filesystem combinations that resulted in fiemap returning some extents that were in practice 0 length on disk (this is possible because fiemap on linux is an advisory output, not necessarily an authoritative one).

So this seems to be fixing a real bug, though I'm uncertain how easy it is to hit - you'd need enough omap to avoid sending them in one push, plus a data allocation pattern in the underlying filesystem that results in empty extents (not sure if this is possible with all-omap objects like bucket indexes or not).

**Comment by satoru-takeuchi** [comment]
@jdurgin Thank you for your input! 

> So this seems to be fixing a real bug, though I'm uncertain how easy it is to hit - you'd need enough omap to avoid sending them in one push, plus a data allocation pattern in the underlying filesystem that results in empty extents (not sure if this is possible with all-omap objects like bucket indexes or not).

Unfortunately, I don't have FIleStore OSDs. So I'd like to know @alcyoneus86 's answer for now. In addition, I'll check the source code to understand your comment in detail (I've not read FileStore's code).

**Pull request closed without merging**